vowpalwabbit/arguments.h

```diff
@@ -0,0 +1,157 @@
+#pragma once
+
+#include <string>
+#include <vector>
+#include <typeinfo>
+#include <memory>
+
+namespace VW {
+
+  struct base_argument {
+    base_argument(std::string name, size_t type_hash)
+      : m_name(name), m_type_hash(type_hash)
+    {}
+
+    std::string m_name;
+    size_t m_type_hash;
+    std::string m_help = "";
+    std::string m_short_name = "";
+    bool m_keep = false;
+
+    virtual ~base_argument() {}
+  };
+
+  template<typename T>
+  struct typed_argument : base_argument {
+    typed_argument(std::string name, T* location)
+      : base_argument(name, typeid(T).hash_code()) {
+      m_locations.push_back(location);
+    }
+
+    static size_t type_hash() {
+      return typeid(T).hash_code();
+    }
+
+    typed_argument& default_value(T value) {
+      m_default_value = std::make_shared<T>(value);
+      return *this;
+    }
+
+    bool default_value_supplied() {
+      return m_default_value.get() != nullptr;
+    }
+
+    T default_value() {
+      return m_default_value ? *m_default_value : T();
+    }
+
+    typed_argument& short_name(std::string short_name) {
+      m_short_name = short_name;
+      return *this;
+    }
+
+    typed_argument& help(std::string help) {
+      m_help = help; return *this;
+    }
+
+    typed_argument& keep(bool keep = true) {
+      m_keep = keep; return *this;
+    }
+
+    bool value_supplied() {
+      return m_value.get() != nullptr;
+    }
+
+    typed_argument& value(T& value) {
+      m_value = std::make_shared<T>(value);
+      return *this;
+    }
+
+    T value(){
+      return m_value ? *m_value : T();
+    }
+
+    std::vector<T*> m_locations;

```

Comment:
Why is this a vector?

Reply:
If there are several places where an argument is specified AND their specification matches then they will both receive the value of the parsed value

Reply:
I'm confused between options and arguments.   The file is named arguments.h, but it seems like this is about specifying options.   Between the two, what do you have in mind here?

Reply:
I decided to use the arguments name because of the arguments object that exists already, so that naming wasn't changed too much. A typed_argument<T> object contains all of the metadata about a specific argument that can be passed on the command line as well as its value once parsed. 

Compared to boost::program_options the typed_argument<T> type maps closely to [option_description](https://www.boost.org/doc/libs/1_63_0/doc/html/boost/program_options/option_description.html)

Reply:
So, by 'argument' you mean 'option', and yes that's my fault.  It might be worth rectifying though since it's cleaner to have 'options' associated with the semantics and 'arguments' associated with the commandline. 

Reparsing, "If there are several places where an option is specified and their specification matches, then when parsing arguments, they will both receive the parsed value".   I understand now, but I'm still a little worried that this is not quite the right structure.  What happens if an option is specified twice, once as a float and once as an int, then an int is passed on the commandline?

Reply:
Sounds good, more than happy to make that naming change.

That is not possible with this implementation, if you attempted to do that it would throw. You can only have multiple locations if everything matches

Reply:
There's a question here about what the right semantics for options are.  Do we want to enforce something about an option across multiple reductions or not?  In the fully modular approach, we would not want to---essentially each reduction's options would live in it's own namespace.   Why should be not be fully modular?

Reply:
It does feel strange to enforce this. The more modular we get here though, the harder it is to decide if an option is unrecognized.

In a fully modular system what do we do if one expects a string and the other an int?

Reply:
To throw on an unrecognized option, you need to know the complete pool of possible options. If you need the complete pool then it becomes impossible to have one option be typed differently based on other options.

Reply:
Sorry, I meant 'throw on an unrecognized argument value'.  Even the current approach allows for parsing with unrecognized options.

Reply:
I see, parsing with unrecognized options is easy. The problem is then being certain about it's final nonrecognition to allow us to throw an error. unrecognized argument value means a value not convertible to the desired type?

Reply:
Yes?  In a library situation where each reduction is a single thing you may only feed the arguments relevant to that reduction.  

But on a commandline, and possibly for convenience in a library situation, it's necessary to have multiple pools of options with each of them parsed against the arguments as an incomplete set.   

What's important is that every argument be parsed by _some_ (at least 1) option.  

To answer my own question, it seems like we should get rid of the vector and just have more options.  All those options in a pool whose argument value are of the correct type (or coercable into the correct type as per int->float) would be parsed and applied with those arguments (and argument values) marked as parsed-at-least-once.  

At the end of parsing from the pool, the present code treats things as an anonymous specification of a data file and barfs if it fails to open.  Something slightly more intelligent may be necessary in general, but we should definitely know about and do something sane (warn or barf) on an unrecognized commandline element.

Reply:
I originally had that behavior essentially, so that every reduction must be very explicit and provide specifications for EVERY option it requires. Knowing when something is unrecognized is still a little tricky but I do think it is possible. I want to make sure we are on the same page here, we can chat about it tomorrow?